src/main/java/net/openhft/chronicle/core/io/VanillaReferenceCounted.java

```diff
@@ -175,8 +186,49 @@ public boolean unmonitored() {
         return unmonitored;
     }
 
+    public void addReferenceChangeListener(ReferenceChangeListener referenceChangeListener) {
+        referenceChangeListeners.add(referenceChangeListener);
+    }
+
+    public void removeReferenceChangeListener(ReferenceChangeListener referenceChangeListener) {
+        referenceChangeListeners.remove(referenceChangeListener);
+    }
+
+    private void notifyListenersAdded(ReferenceOwner referenceOwner) {
+        this.callReferenceChangeListeners(
+                (listener, referenceCounted, lhs, rhs) -> listener.onReferenceAdded(referenceCounted, lhs),
+                referenceOwner, null);
+    }
+
+    private void notifyListenersRemoved(ReferenceOwner referenceOwner) {
+        this.callReferenceChangeListeners(
+                (listener, referenceCounted, lhs, rhs) -> listener.onReferenceRemoved(referenceCounted, lhs),
+                referenceOwner, null);
+    }
+
+    private void notifyListenersTransferred(ReferenceOwner from, ReferenceOwner to) {
+        this.callReferenceChangeListeners(ReferenceChangeListener::onReferenceTransferred, from, to);
+    }
+
+    private void callReferenceChangeListeners(ListenerInvoker listenerInvoker, ReferenceOwner lhs, ReferenceOwner rhs) {
+        //noinspection ForLoopReplaceableByForEach
+        for (int i = 0; i < referenceChangeListeners.size(); i++) {
+            ReferenceChangeListener listener;
+            try {
+                listener = referenceChangeListeners.get(i);
+            } catch (ArrayIndexOutOfBoundsException e) {
+                // This can happen if a listener is removed concurrently during the iteration
+                continue;
+            }
+            listenerInvoker.invokeListener(listener, this, lhs, rhs);
+        }
+    }
+
     private ClosedIllegalStateException newReleasedClosedIllegalStateException() {
         return new ClosedIllegalStateException(type.getName() + " released");
     }
 
+    private interface ListenerInvoker {
+        void invokeListener(ReferenceChangeListener listener, ReferenceCounted referenceCounted, ReferenceOwner lhs, ReferenceOwner rhs);
+    }
```

Comment:
Having this interface means we can create non-capturing lambdas in `notifyListenersXXX` so we can re-use `callReferenceChangeListeners` for adds, removes and transfers